src/libsyntax/codemap.rs

```diff
@@ -129,6 +132,15 @@ pub struct Span {
     pub expn_id: ExpnId
 }
 
+/// Spans are converted to MultiSpans just before error reporting, either automatically,
+/// generated by line grouping, or manually constructed.
+/// In the latter case care should be taken to ensure that spans are ordered, disjoint,
+/// and point into the same FileMap.
+#[derive(Clone)]
+pub struct MultiSpan {
+    pub spans: SmallVector<Span>
```

Comment:
Using `SmallVector` here seemed nice at first since
- converting a Span into a `MultiSpan` is cheap
- `const COMMAND_LINE_MSP` can exist and be used inside patterns (well, after the next snapshot -> #30117)

Since this only matters once when we actually want to report something this should probably be `Vec<Span>`?